pkg/discovery/worker/garbage_collector.go

```diff
@@ -0,0 +1,94 @@
+package worker
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+
+	api "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/turbonomic/kubeturbo/pkg/action/executor"
+)
+
+type GarbageCollector struct {
+	client                *kubernetes.Clientset
+	finishCollecting      chan bool
+	collectionIntervalMin int
+
+	// Necessary for test
+	podAge time.Duration
+}
+
+func NewGarbageCollector(client *kubernetes.Clientset, finishChan chan bool, collectionIntervalMin int, podAge time.Duration) *GarbageCollector {
+	return &GarbageCollector{
+		podAge:                podAge,
+		client:                client,
+		finishCollecting:      finishChan,
+		collectionIntervalMin: collectionIntervalMin,
+	}
+}
+
+func (g *GarbageCollector) StartCleanup() {
+	glog.V(4).Info("Start leaked pods cleanup.")
+	// Also cleanup immediately at startup
+	g.cleanupLeakedPods()
+	go func() {
+		collectionInterval := time.Duration(g.collectionIntervalMin) * time.Minute
+		ticker := time.NewTicker(collectionInterval)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-g.finishCollecting:
+				// This would happen when kubeturbo is exiting
+				return
+			case <-ticker.C:
+				g.cleanupLeakedPods()
+			}
+		}
+	}()
+}
+
+func (g *GarbageCollector) cleanupLeakedPods() {
+	// Get all pods which have the gc annotation. This can include those on which an
+	// action is being executed right now.
+	listOpts := metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=%s", executor.TurboGCLabelKey, executor.TurboGCLabelVal)}
+	podList, err := g.client.CoreV1().Pods("").List(context.TODO(), listOpts)
+	if err != nil {
+		glog.Warningf("Error getting leaked pods: %v", err)
```

Comment:
Should this be returned just to be safe?  Is the `List` function always return nil if there is an error?

Reply:
As far as I know list is always nil on error. As you said to be safe on error have added a `return`.